File: src/com/urbanairship/octobot/TaskExecutor.scala
package com.urbanairship.octobot

import java.util.HashMap
import java.lang.reflect.Method
import org.json.JSONObject

object TaskExecutor {

  val taskCache = new HashMap[String, Method]

  def execute(taskName: String, message: JSONObject) {
    var method: Method = null

    if (taskCache.containsKey(taskName)) {
      method = taskCache.get(taskName)
    } else {
      val task = Class.forName(taskName)
      val klass = new JSONObject().getClass
      method = task.getMethod("run", klass)
      taskCache.put(taskName, method)
    }
    
    method.invoke(null, message)
  }

}

File: src/main/java/com/urbanairship/octobot/TaskExecutor.java
package com.urbanairship.octobot;

import java.util.HashMap;
import java.lang.reflect.Method;
import org.json.simple.JSONObject;
import java.lang.reflect.InvocationTargetException;

public class TaskExecutor {

    private static final HashMap<String, Method> taskCache =
            new HashMap<String, Method>();

    @SuppressWarnings("unchecked")
    public static void execute(String taskName, JSONObject message) 
            throws ClassNotFoundException,
                   NoSuchMethodException,
                   IllegalAccessException,
                   InvocationTargetException {

        Method method = null;

        if (taskCache.containsKey(taskName)) {
            method = taskCache.get(taskName);
        } else {
            Class task = Class.forName(taskName);
            method = task.getMethod("run", new Class[]{ JSONObject.class });
            taskCache.put(taskName, method);
        }

        method.invoke(null, new Object[]{ message });
    }

}
